opengrok-indexer/src/test/java/org/opengrok/indexer/util/HeadHandlerTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
+ */
+
+package org.opengrok.indexer.util;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class HeadHandlerTest {
+    private class RandomInputStream extends InputStream {
+        private final int maxcount;
+        private int count;
+        private final int maxlines;
+        private int lines;
+
+        private String letters;
+
+        public RandomInputStream(int count, int lines) {
+            this.maxcount = count;
+            this.count = 0;
+            this.maxlines = lines;
+            this.lines = 0;
+
+            letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+            letters = letters + letters.toLowerCase();
+            letters = letters + "0123456789";
+        }
+
+        public int getCount() {
+            return count;
+        }
+
+        @Override
+        public int read() throws IOException {
+            if (++count < maxcount) {
+                // Want the newlines to appear within the first half.
+                if (count % ((maxcount / 2) / maxlines) == 0) {
+                    return '\n';
+                } else {
+                    return letters.charAt((int) (Math.random() % letters.length()));
+                }
+            }
+
+            return -1;
+        }
+    }
+
+    @Test
+    public void testHeadHandler() throws IOException {
+        final int lines = 5;
+        final int headLines = 3;
+        int totalCount = 8192;
+
+        RandomInputStream rndStream = new RandomInputStream(totalCount, lines);
+        HeadHandler handler = new HeadHandler(headLines);
+        assertTrue(totalCount > handler.getBufferedReaderSize());
+        handler.processStream(rndStream);
+        assertEquals(headLines, handler.size());
+        assertEquals(totalCount + 1, rndStream.getCount());
+        int linesLen = 0;
+        for (int i = 0; i < headLines; i++) {
+            linesLen += handler.get(i).length();
```

Comment:
With this approach using randomness, I wonder if it would be informative to compare the line content to what was generated.

Reply:
For me it seems a little much work for a test; I'd prepare some files as resources (perhaps with this generator) and assert the content.

Reply:
I wanted the test to adapt automatically based on changes to `HeadHandler` - in particular the internal buffer size.